test_runner/batch_others/test_read_validation.py

```diff
@@ -0,0 +1,183 @@
+from contextlib import closing
+
+from fixtures.zenith_fixtures import ZenithEnv
+from fixtures.log_helper import log
+
+from psycopg2.errors import UndefinedTable
+from psycopg2.errors import IoError
+
+pytest_plugins = ("fixtures.zenith_fixtures")
+
+extensions = ["pageinspect", "zenith_test_utils", "pg_buffercache"]
+
+
+#
+# Validation of reading different page versions
+#
+def test_read_validation(zenith_simple_env: ZenithEnv):
+    env = zenith_simple_env
+    env.zenith_cli.create_branch("test_read_validation", "empty")
+
+    pg = env.postgres.create_start("test_read_validation")
+    log.info("postgres is running on 'test_read_validation' branch")
+
+    with closing(pg.connect()) as con:
+        with con.cursor() as c:
+
+            for e in extensions:
+                c.execute("create extension if not exists {};".format(e))
+
+            c.execute("create table foo (c int) with (autovacuum_enabled = false)")
+            c.execute("insert into foo values (1)")
+
+            c.execute("select lsn, lower, upper from page_header(get_raw_page('foo', 'main', 0));")
+            first = c.fetchone()
+
+            c.execute("select relfilenode from pg_class where relname = 'foo'")
+            relfilenode = c.fetchone()[0]
+
+            c.execute("insert into foo values (2);")
+            c.execute("select lsn, lower, upper from page_header(get_raw_page('foo', 'main', 0));")
+            second = c.fetchone()
+
+            assert first != second, "Failed to update page"
+
+            log.info("Test table is populated, validating buffer cache")
+
+            c.execute(
+                "select count(*) from pg_buffercache where relfilenode =  {}".format(relfilenode))
+            assert c.fetchone()[0] > 0, "No buffers cached for the test relation"
+
+            c.execute(
+                "select reltablespace, reldatabase, relfilenode from pg_buffercache where relfilenode = {}"
+                .format(relfilenode))
+            reln = c.fetchone()
```

Comment:
seems a bit odd to get 'reltablespace' and 'reldatabase' from pg_buffercache. Would seem more straightforward to get them directly in the query from pg_class above, with something like:

```
select reltablespace,
  (select oid from pg_database where datname = current_database()),
  relfilenode
from pg_class where relname='foo'
```

Reply:
I agree that is a roundabout way to getting reltablespace, but pg_class only carries object's table space when it doesn't match database's default table space (it's 0 if default is used) this makes the query quite a bit more complicated
 
```
select c.relfilenode, d.oid reldatabase,
       case when c.reltablespace = 0 then d.dattablespace 
            else c.reltablespace end reldatabase 
from pg_class c, pg_database d 
where relname = 'foo' and datname = current_database();
```